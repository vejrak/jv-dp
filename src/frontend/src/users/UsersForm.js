// @flow

import React from 'react'
import { Button } from 'react-bootstrap'
import Select from 'react-select'
import type { ErrorResponse, Group, User } from '../types'
import { createUser } from './actions'
import { getApprovedRoles } from './helpers'
import Spinner from '../components/Spinner'
import * as validation from '../helpers/validation'
import Input from '../components/Input'
import { isFormValid, createGroupsOptions } from '../helpers'
import { getGroups } from '../group/actions'

type Props = $ReadOnly<{|
  createUser: typeof createUser,
  isFetching: boolean,
  user: ?User,
  roles: Object,
  groups: Array<Group>,
  getGroups: typeof getGroups,
  error: ?ErrorResponse,
|}>

type State = {
  formData: {
    username: string,
    password: string,
    rank: number,
    groups: Array<Object>,
  },
  validationResults: validation.ValidationResults<*>,
}

class UsersForm extends React.PureComponent<Props, State> {
  state = {
    formData: { username: '', password: '', rank: 3, groups: [] },
    validationResults: {},
  }

  componentDidMount() {
    this.props.getGroups()
  }

  validate = () => {
    const { formData } = this.state
    return {
      username: validation.required(formData.username),
      password: validation.required(formData.password),
    }
  }

  handleChange = (value: Object): void => {
    const { formData, validationResults } = this.state
    delete validationResults[Object.keys(value)[0]]
    this.setState({
      formData: { ...formData, ...value },
      validationResults,
    })
  }

  resetState = (): void => {
    this.setState({
      formData: { username: '', password: '', rank: 3, groups: [] },
      validationResults: {},
    })
  }

  handleClick = async () => {
    const { formData } = this.state
    const validationResults = this.validate()
    this.setState({ validationResults })
    if (isFormValid(validationResults)) {
      await this.props.createUser({
        ...formData,
        groups: formData.groups.map((value) => value.value),
      })
      this.resetState()
    }
  }

  render() {
    const { isFetching, error, groups, roles, user } = this.props
    const { formData, validationResults } = this.state
    const options = createGroupsOptions(groups)

    if (isFetching) return <Spinner />

    const approvedRoles = getApprovedRoles(1, roles)

    return (
      <div className="form-group p-3">
        {user && (
          <div className="alert alert-success" role="alert">
            User {user.username} created !
          </div>
        )}
        {error && (
          <div className="alert alert-danger" role="alert">
            {error.errmsg}
          </div>
        )}
        <label className="mt-2" htmlFor="username">
          Username
        </label>
        <Input
          className="form-control"
          id="username"
          name="username"
          onChange={({ target }) =>
            this.handleChange({ username: target.value })
          }
          validationError={validationResults.username}
          value={formData.username}
        />
        <label className="mt-2" htmlFor="password">
          Password
        </label>
        <Input
          className="form-control"
          id="password"
          name="password"
          onChange={({ target }) =>
            this.handleChange({ password: target.value })
          }
          type="password"
          validationError={validationResults.password}
          value={formData.password}
        />
        <label className="mt-2" htmlFor="sel1">
          Roles
        </label>
        <select
          className="form-control"
          id="sel1"
          onChange={({ target }) =>
            this.handleChange({ rank: parseInt(target.value, 10) })
          }
          value={formData.rank}
        >
          {Object.keys(approvedRoles).map((index) => (
            <option key={approvedRoles[index]} value={approvedRoles[index]}>
              {index}
            </option>
          ))}
        </select>
        <label className="mt-2" htmlFor="groups">
          Groups
        </label>
        <Select
          className="mt-2"
          id="groups"
          isMulti
          onChange={(value) => this.handleChange({ groups: value })}
          options={options}
          value={formData.groups}
        />
        <Button
          className="mt-3 btn btn-success d-block"
          disabled={isFetching}
          onClick={this.handleClick}
        >
          Create
        </Button>
      </div>
    )
  }
}

export default UsersForm
