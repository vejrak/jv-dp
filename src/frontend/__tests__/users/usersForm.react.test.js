import React from 'react'
import renderer from 'react-test-renderer'
import UsersForm from '../../src/users/UsersForm'

const mockProps = {
  isFetching: false,
  error: null,
  roles: { SUPER_ADMIN: 1, ADMIN: 2, USER: 3 },
  user: null,
  createUser: () => {},
  getUsers: () => {},
  getGroups: () => {},
}

it('renders correctly', () => {
  const tree = renderer.create(<UsersForm {...mockProps} />).toJSON()
  expect(tree).toMatchSnapshot()
})
