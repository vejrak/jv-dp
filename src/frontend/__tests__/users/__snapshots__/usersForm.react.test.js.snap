// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="form-group p-3"
>
  <label
    className="mt-2"
    htmlFor="username"
  >
    Username
  </label>
  <div>
    <input
      className="form-control"
      id="username"
      name="username"
      onChange={[Function]}
      value=""
    />
  </div>
  <label
    className="mt-2"
    htmlFor="password"
  >
    Password
  </label>
  <div>
    <input
      className="form-control"
      id="password"
      name="password"
      onChange={[Function]}
      type="password"
      value=""
    />
  </div>
  <label
    className="mt-2"
    htmlFor="sel1"
  >
    Roles
  </label>
  <select
    className="form-control"
    id="sel1"
    onChange={[Function]}
    value={3}
  >
    <option
      value={2}
    >
      ADMIN
    </option>
    <option
      value={3}
    >
      USER
    </option>
  </select>
  <label
    className="mt-2"
    htmlFor="groups"
  >
    Groups
  </label>
  <div
    className="css-1pcexqc-container mt-2"
    id="groups"
    onKeyDown={[Function]}
  >
    <div
      className="css-bg1rzq-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="css-1hwfws3"
      >
        <div
          className="css-151xaom-placeholder"
        >
          Select...
        </div>
        <div
          className="css-1g6gooi"
        >
          <div
            className=""
            style={
              Object {
                "display": "inline-block",
              }
            }
          >
            <input
              aria-autocomplete="list"
              autoCapitalize="none"
              autoComplete="off"
              autoCorrect="off"
              disabled={false}
              id="react-select-2-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              spellCheck="false"
              style={
                Object {
                  "background": 0,
                  "border": 0,
                  "boxSizing": "content-box",
                  "color": "inherit",
                  "fontSize": "inherit",
                  "label": "input",
                  "opacity": 1,
                  "outline": 0,
                  "padding": 0,
                  "width": "1px",
                }
              }
              tabIndex="0"
              type="text"
              value=""
            />
            <div
              style={
                Object {
                  "height": 0,
                  "left": 0,
                  "overflow": "scroll",
                  "position": "absolute",
                  "top": 0,
                  "visibility": "hidden",
                  "whiteSpace": "pre",
                }
              }
            >
              
            </div>
          </div>
        </div>
      </div>
      <div
        className="css-1wy0on6"
      >
        <span
          className="css-bgvzuu-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          className="css-16pqwjk-indicatorContainer"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <svg
            aria-hidden="true"
            className="css-19bqh2r"
            focusable="false"
            height={20}
            viewBox="0 0 20 20"
            width={20}
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <button
    className="mt-3 btn btn-success d-block btn btn-primary"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Create
  </button>
</div>
`;
